### ISSUE TITLE
UI Laggs when retrieving results from /refresh

### ISSUE THREAD

**Issue by chrisknoll** [OHDSI/Atlas#2820]
### Expected behavior
UI should be responsive when clicking

### Actual behavior
The permission configuration we have is leading to a huge `permissions` payload when fetched from the server with /refresh.  See attached for reference. The size of the payload is 257KB as seen in this screenshot:

![image](https://user-images.githubusercontent.com/6818777/215250192-c09abe26-b50a-4fac-95b5-4fe37c263133.png)

### Steps to reproduce behavior

Our environment may be unique: we have 20-30 active CDM sources configured in webapi.  However, we don't adopt any special groups or role permissions.   We only leverage source-level permissions to generate and view results which should be fairly standard practice for security in atlas.

Note, opening the attached file (256KB) in notepad++ makes the app hang (probably trying to parse something in the file).

[refreshPayload.txt](https://github.com/OHDSI/Atlas/files/10526082/refreshPayload.txt)

**Comment by anton-abushkevich** [comment]
On my system your permissions file are recieved/parsed in about 200ms-700ms, with no visible lag. 

Notepad++ is indeed freezing for several seconds when opening the file. It is definitely trying to parse something. Regular windows notepad, however, opens the file without any problem.

**Comment by chrisknoll** [comment]
Right, but the problem is that the payload is taking several seconds to load into the browser (some users with 350+KB payloads say they see 'waiting for browser' prompts in their browser, its bad.

It's like if you were using VSCode and every time you hit a tab, or a menu option or a hotkey, you had to download and parse 350KB of data from Microsoft.com.   We shouldn't be doing this operation on every user intraction on the site.  I think the reason for the call is that we want to get the latest permissions. However, I think that we should only be loading permissions 2 times:  once when you log in, and once at a point you get a 403: unauthorized response from the server.   This way, we only need to notify users that some page is unavailable (ie: when they try to do something they aren't allowed to do), which is very rare because user permissions rarely change during their user session in the app.

**Comment by chrisknoll** [comment]
To give some context to the `/refresh` spam, this is the capture in atlas when I just clicked through the left nav and did a serach.   It refreshed permissions 10 times in 30 seconds.   I'm not sure what is causing the UI lag, but it definitely lags the UI while the permissions are reloading (the user icon in the top right turns into a  'swirl' as it tries to figure out your login).   I can try to dig into the browser profiler to identify where the bottleneck is, but I think the primary problem is that we don't need to spam refresh on permissions every time we click a button.

**Comment by anton-abushkevich** [comment]
Yes, I agree, these /refresh calls is very much redundant and require reworking. Permissions are there since 2018, see [org.ohdsi.webapi.shiro.filters.UpdateAccessTokenFilter#149](https://github.com/OHDSI/WebAPI/blob/master/src/main/java/org/ohdsi/webapi/shiro/filters/UpdateAccessTokenFilter.java#L149). The filter was initailly only for token refresh, and permissions were added to the response later.

**Comment by chrisknoll** [comment]
yes, understood. We haven't had much time to experience this behavior because we only enabled security in our own environment in the past year.   But putting asde this is long standing behavior and we've not enabled it long:

Looking at the code, it seems at one point, the JWT was passed permissions in `createJsonWebToken()` but the current implementation assigns the value via `requset.setAttribute(PERMISSIONS_ATTRIBUTE...)`.

Also, looking at the purpose of the /refresh end-piont, I understand that the JWT needs to be 'refreshed' to avoid login timeout.  So, IMO, refresh was overloaded to include permission information instead of just refreshing the JWT.  These two lines in `UpdateAccessTokenFilter.prehandle()` are not related (or necessary) for updating the JWT.

```
    Collection<String> permissions = this.authorizer.getAuthorizationInfo(login).getStringPermissions();
    request.setAttribute(PERMISSIONS_ATTRIBUTE, StringUtils.join(permissions, "|"));
```

Both classes: `SendTokenInHeaderFilter` and `UpdateAccessTokenFilter` are specifically about the JWT, and the permission information was a bolt-on.

My suggestion is that we think about separating JWT token concerns from permission concerns since permission is such a heavier payload and requires much more client-side processing (JWT token is simply stored, while permissions must be parsed).  I understand why we'd inject a JWT into the headers of the HTTP requests, but I do not understand the need to do this for permissions since you would never want a client to inject their own permissions into a HTTP request (the server should always look for its own permission records, not an untrusted client).   

A simple question is:   what would happen if we removed the permission attribute from the UpdateAccessTokenFilter?

**Comment by chrisknoll** [comment]
Updated Information:

Patrick had a problem as he was navigating around Atlas:

> I have noticed that when I have an ATLAS session open and I continue to do a lot of clicking between analyses and sources, that the more I do that, the worse the user authentication problem appears to get.  If I close Chrome and restart ATLAS, it starts good and then cumulatively gets worse.  I’ve also noticed watching my Task Manager that memory use by Chrome seems to grow as I click through ATLAS over a session.   So it is possible that theres some process that’s growing the authentication instead of replacing?  Or some other memory leak associated with it?

@anton-abushkevich : you mentioned you didn't see this behavior, but did you navigate around the app long enough to see an impact?  It's possible what Patrick sees is some sort of memory leak, and wanted to ask you to try to profile the app if you notice any memory consumption or anything that would indicate an increasing load on the browser over time.

I still believe that the solution is to separate the refresh call to just refresh the JWT token, and provide a separate endpoint for permissions.

**Comment by chrisknoll** [comment]
I clicked around the app and started to see the slowdown, and to try to capture the 'simplest' navigation, i went from 'about' to 'home' and recorded this activity in chrome:

![image](https://user-images.githubusercontent.com/6818777/218616725-07517660-979a-4ad6-9abc-6e194b4a98c7.png)

Thi sis showing about 2 seconds of execution time surrounding many calls to 'isPermitted'.   One surprising thing about this is why it is checking for permissions against sources and cohort definitions when the active tab isn't even cohort definitions.

**Comment by anthonysena** [comment]
From discussion on today's Atlas developer call, we think that this issue may stem from the code here:

https://github.com/OHDSI/Atlas/blob/190dfee949bfd5a2966f74f0ab64a5e2c71b7f05/js/pages/Route.js#L46-L59

We are suspicious of this code since it is presumably creating a subscription each time a new Route is requested in the application. We also did not see where this specific subscription is then disposed which may cause a massive cascade of browser notifications which could be the cause of the UI lag. More investigation is required as this may not be the root cause.

**Comment by chrisknoll** [comment]
Additional Info:

I set breakpoints on the subscription, and it doesn't appear to create multiple subscriptions.  It doesn't mean that there isn't a leak somewhere, but at least I don't see evidence of subscriptions being doubled up.

I did trace through the code to understand how tokens were being processed.   A few questions / concerns:

In Route.js checkPermission():

```
		checkPermission() {
			if (authApi.authProvider() === authApi.AUTH_PROVIDERS.IAP) {
				return authApi.loadUserInfo();
			} else if (appConfig.userAuthenticationEnabled && authApi.token() != null && authApi.tokenExpirationDate() > **new Date()**) {
				return authApi.refreshToken();
			}
			return Promise.resolve();
		}
```

What is strange is the date check:   if the token expiration date is in the future, why would we want to refresh the token?

**Comment by anthonysena** [comment]
Based on discussion in the Atlas WG, for now we'll move this to v2.14 so we can have further discussion around solutioning here.
